xapian-core/tests/api_wrdb.cc

```diff
@@ -1589,6 +1589,41 @@ DEFINE_TESTCASE(nomoredocids1, writable) {
     return true;
 }
 
+// Test synonyms from subdbs are appropriately merged.
+DEFINE_TESTCASE(synonym_merge1, writable && synonyms) {
+    Xapian::WritableDatabase db = get_writable_database();
+    Xapian::WritableDatabase db1 = get_named_writable_database("db1");
+
+    db.add_synonym("abc", "asd");
+    db.add_synonym("abc", "asd2");
+    db.add_synonym("abc", "asd4");
+    db.add_synonym("abc", "asd6");
+    db.commit();
+
+    db1.add_synonym("abc", "asd1");
```

Comment:
I'll add `asd2` to this shard too so we test that merging correctly handles the same synonym coming from more than one shard.